### PR TITLE
- Avoid going via a temporary IdealNodesList.

### PR DESCRIPTION
**Pull request by baldersheim**
- Use ConstArrayRef to hide implementation.
- Store all 3 node categories in a single vector.
- Use a small_vector that can handle redundancy up to 5 without requiring extra memory allocation.
- Build a hash_map if redundancy/groups > 32 for constant lookup time.

@vekterli PR
@geirst FYI

### CODE REVIEW

**Review by vekterli — DISMISSED** [review]
:+1: the ideal state caching is becoming rather complex... At least now that we have an optimized order-lookup mechanism, we should look for more places where it can be used.

File: storage/src/vespa/storage/distributor/ideal_service_layer_nodes_bundle.h [hunk 1]
```diff
@@ -1,38 +1,75 @@
...
 /*
  * Bundle of ideal service layer nodes for a bucket.
  */
 class IdealServiceLayerNodesBundle {
-    std::vector<uint16_t> _available_nodes;
-    std::vector<uint16_t> _available_nonretired_nodes;
-    std::vector<uint16_t> _available_nonretired_or_maintenance_nodes;
-    vespalib::hash_set<uint16_t> _unordered_nonretired_or_maintenance_nodes;
 public:
+    using ConstNodesRef = vespalib::ConstArrayRef<uint16_t>;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/ideal_service_layer_nodes_bundle.h:13]: ConstArrayRef understands a SmallVector, as it understands std::vector++ ....
I will keep in mind and see what we can do to use std::span

File: storage/src/vespa/storage/distributor/statecheckers.cpp [hunk 2]
```diff
@@ -484,11 +492,7 @@ SplitInconsistentStateChecker::getReason(const document::BucketId& bucketId, con
 
 namespace {
 
-bool
-isInconsistentlySplit(const StateChecker::Context& c)
-{
-    return (c.entries.size() > 1);
-}
+
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/statecheckers.cpp:495]: Fixed

File: storage/src/tests/distributor/statecheckerstest.cpp [hunk 3]
```diff
@@ -1587,7 +1588,8 @@ TEST_F(StateCheckersTest, context_populates_ideal_state_containers) {
     StateChecker::Context c(node_context(), operation_context(),
                             getDistributorBucketSpace(), statsTracker, makeDocumentBucket({17, 0}));
 
-    ASSERT_THAT(c.idealState(), ElementsAre(1, 3));
+    ASSERT_EQ(1, c.idealState()[0]);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/tests/distributor/statecheckerstest.cpp:1592]: Fixed

File: storage/src/vespa/storage/storageutil/utils.h [hunk 4]
```diff
@@ -66,7 +66,7 @@ std::string dumpVector(const std::vector<A>& vec) {
 }
 
 template<class A>
-bool hasItem(const std::vector<A>& vec, A entry) {
+bool hasItem(vespalib::ConstArrayRef<A> vec, A entry) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/storageutil/utils.h]: Good point, fixed.

File: storage/src/vespa/storage/distributor/ideal_service_layer_nodes_bundle.h [hunk 5]
```diff
@@ -1,38 +1,75 @@
...
-    const std::vector<uint16_t> & available_nodes() const noexcept { return _available_nodes; }
-    const std::vector<uint16_t> & available_nonretired_nodes() const noexcept { return _available_nonretired_nodes; }
-    const std::vector<uint16_t> & available_nonretired_or_maintenance_nodes() const noexcept {
-        return _available_nonretired_or_maintenance_nodes;
+    NonRetiredOrMaintenance2Index nonRetiredOrMaintenance2Index() const noexcept {
+        return NonRetiredOrMaintenance2Index(*this);
     }
-    bool is_nonretired_or_maintenance(uint16_t node) const noexcept {
-        return _unordered_nonretired_or_maintenance_nodes.contains(node);
+    ConstNodesRef2Index available2Index() const noexcept {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/ideal_service_layer_nodes_bundle.h]: Fixed

File: storage/src/vespa/storage/distributor/operationtargetresolverimpl.h [hunk 6]
```diff
@@ -59,13 +58,11 @@ class BucketInstanceList : public vespalib::AsciiPrintable {
      *   _instances.size() >= configured redundancy level, unless insufficient
      *   number of nodes are available
      */
-    void extendToEnoughCopies(const DistributorBucketSpace& distributor_bucket_space,
-                              const BucketDatabase& db,
-                              const document::BucketId& targetIfNonPreExisting,
-                              const document::BucketId& mostSpecificId);
+    void extendToEnoughCopies(const DistributorBucketSpace& distributor_bucket_space, const BucketDatabase& db,
+                              const document::BucketId& targetIfNonPreExisting, const document::BucketId& mostSpecificId);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operationtargetresolverimpl.h]: Yes, that is a matter of taste. Long parameter lists are often a bad sign anyway. I favor wider over many lines, with the 120c/line hint.
Leaving as it was.

**Comment by baldersheim** [comment]
I have now addressed all comments actionable. Will merge once green.

**Pull request merged**